### PR TITLE
Add rules for hybrid Node.js and browser environments

### PR DESCRIPTION
**Pull request by Mrtenz**
This adds a new package for hybrid Node.js and browser envrionments. It prevents the use of any Node.js or browser-specific globals.

Because we make use of Node.js and browser APIs in many modules right now, adding this to an existing package would cause a lot of breaking changes. That's why I created a new package instead. It includes a script to update the rules automatically.

Using the following file as example:

```ts
import { createHash } from 'crypto';
import { join } from 'path';

export const doThing = () => {
  const value = Buffer.from('foo bar');
  createHash('sha256').update(value).digest('hex');
  join('foo');
};
```

Results in these errors:

```
  1:1   error  'crypto' import is restricted from being used. Use 'ethereum-cryptography' instead        no-restricted-imports
  2:1   error  'path' import is restricted from being used. This module is not available in the browser  no-restricted-imports
  5:17  error  Unexpected use of 'Buffer'. Use 'Uint8Array' instead                                      no-restricted-globals
```

## To do

- [ ] Disable rules in test files?
- [x] Test in production to ensure it works as expected.

### CODE REVIEW

**Comment by Mrtenz** [comment]
I'm not sure what the best way would be to handle test files. For testing in production, I simply added this package to the `*.ts` override, which would also apply it to test files. We could set it up in the module template to turn off those rules for `*.test.ts` maybe.

**Comment by Gudahtt** [comment]
Could we add this to the base package instead? That package is already meant to prevent the use of anything Node.js or browser-specific.

I'm still confused as to why Node.js specific globals aren't already disallowed though :thinking: The base package uses `no-undef`, and doesn't use an environment that includes globals like Buffer.

**Comment by Gudahtt** [comment]
I think this is the culprit: https://github.com/typescript-eslint/typescript-eslint/blob/efc147b04dce52ab17415b6a4ae4076b944b9036/packages/eslint-plugin/src/configs/eslint-recommended.ts

It's disabled because TypeScript is already checking for undefined types. But TypeScript isn't seeing it as undefined because we have `@types/node` installed (presumably so that we can use Node.js APIs in tests?). Not sure what the best way to solve this would be - either explicitly re-add `no-undef` in our TypeScript ESLint config, or ask TypeScript to not include `@types/node` when checking the production files by default.

**Comment by Gudahtt** [comment]
>I'm afraid that it will cause many breaking changes in projects, preventing them from upgrading to the latest config. Fixing these errors is not as simple as changing one line of code, some possibly requiring a larger refactor (like swapping out Buffer for Uint8Array and crypto with a browser-friendly version).

In the past, we've handled this by disabling disruptive rules in the project ESLint config to make the update easier. Then we can re-enable any rules one-by-one to fix lint errors. We're made many disruptive lint changes recently, and many more are planned (e.g. enabling the ESLint TypeScript rules that require type information); making a new package each time would be confusing.

**Comment by Gudahtt** [comment]
>Not sure what the best way to solve this would be - either explicitly re-add `no-undef` in our TypeScript ESLint config, or ask TypeScript to not include `@types/node` when checking the production files by default.

I looked into this, and couldn't find a reasonable way of excluding `@types/node` from `tsconfig.build.json` specifically, while including it in `tsconfig.json`.

So I guess we should explicitly add no-undef to our TypeScript ESLint config.

File: packages/browser/scripts/generate-rules.js [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+const generateRules = () => {
+  return {
+    // Disallow use of Node.js built-in modules.
+    'no-restricted-imports': [
+      'error',
+      {
+        paths: builtinModules.map(getModuleRule).flat(),
+      },
+    ],
+    'no-restricted-globals': ['error', ...getGlobals()],
```
Comment (Gudahtt) [inline packages/browser/scripts/generate-rules.js:85]: If we agree on adding `no-undef` to our TypeScript config, then I don't think we need this rule anymore. The ESLint environment should be enough to explain what globals exist in that environment; we only need this rule if we want to include a global in the environment but still prevent it from being used.

File: packages/browser/scripts/generate-rules.js [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+      message:
+        MESSAGE_OVERRIDES[name] ||
+        'This global is not available in both the browser and Node.js.',
+    }));
+};
+
+const generateRules = () => {
+  return {
+    // Disallow use of Node.js built-in modules.
+    'no-restricted-imports': [
```
Comment (Gudahtt) [inline packages/browser/scripts/generate-rules.js:79]: Would this work instead? https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/no-nodejs-modules.md

File: packages/browser/scripts/generate-rules.js (hunk 1 above)
Comment (Mrtenz) [reply packages/browser/scripts/generate-rules.js:85]: Yeah, that seems like a better solution, and would also remove the need for rule generation combined with the other proposed rule.

Comment (Mrtenz) [reply packages/browser/scripts/generate-rules.js]: Apparently `no-undef` also breaks stuff that is available in both Node.js and browsers.

```
'console' is not defined.
```

Comment (Gudahtt) [reply packages/browser/scripts/generate-rules.js]: Hmm. That should be brought in by the `shared-node-browser` environment. Maybe the base config environment is getting overwritten in the typescript config? If so we could re-add it.

Comment (Mrtenz) [reply packages/browser/scripts/generate-rules.js]: Oh, right. I just had to add an environment to the `.eslintrc.js` we use to lint this package itself. 😄

**Review by FrederikBolding — APPROVED** [review]

File: packages/browser/scripts/generate-rules.js (hunk 1 above)
Comment (Gudahtt) [reply packages/browser/scripts/generate-rules.js]: Where were you seeing this error? Locally I am not seeing it, even after removing the environment.

I thought using `extend` would bring the environments along, so they wouldn't need to be re-declared.

Comment (Gudahtt) [reply packages/browser/scripts/generate-rules.js]: Removed here: https://github.com/MetaMask/eslint-config/pull/255

Comment (Mrtenz) [reply packages/browser/scripts/generate-rules.js]: Might have been just the test that needs it. The test doesn't use any config other than the one being tested.

**Pull request merged**